modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +76,47 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): F[Set[P2PContext]] = for {
+    _ <- twoWayHandshake(context, none).attempt.map(_.left.map { err =>
+      logger.error(err)(s"Joining to P2PContext ${context.show} failed")
+    })
+    maybePeer <- clusterStorage.getPeer(context.id)
+    maybeDiscovered = maybePeer.map {
+      peerDiscovery
+        .discoverFrom(_)
+        .map(_.map(toP2PContext))
+        .handleErrorWith { err =>
+          logger
+            .error(err)(s"Peer discovery from P2PContext ${context.show} failed")
+            .as(Set.empty)
+        }
+    }
+    discovered <- maybeDiscovered.getOrElse(Set.empty[P2PContext].pure)
+  } yield discovered
+
+  private def joinAll(toJoin: PeerToJoin): F[Unit] = {
+    type Agg = (Set[P2PContext], Set[P2PContext])
+    type Res = Unit
+
+    (Set(toJoin): Set[P2PContext], Set.empty[P2PContext]).tailRecM {
+      case (joining, _) if joining.isEmpty =>
+        ().asRight[Agg].pure
+      case (joining, joined) =>
+        for {
+          discovered <- joining.toList.parTraverse(joinAndDiscover)
+          updatedJoined = joined ++ joining
+          updatedJoining = discovered.flatten.filterNot(joining.contains).toSet
```

Comment:
Several points:

- I would investigate a bit more how does the discovery work internally to decide whether we need to filter out the nodes here additionally.
- Another thing if my intuition is right you should use `updatedJoined` to filter out discovered peers and not `joining` if filtering is needed.
- Another thing is that not all peers in `joining` will always be joined, so the second parameter called `joined` here should rather be called `attempted`.
- I would also think about maybe moving discovery to the `Joining` class and perform discovery based on the known details in the `joinAll` (the attempted, failed and joined nodes) and only offload the last discovered to the ref that would be available to fetch from.

Reply:
If you need to filter out in this line maybe that syntax would be better:
`updatedJoining = discovered.flatten.toSet -- joining`

Reply:
> * I would investigate a bit more how does the discovery work internally to decide whether we need to filter out the nodes here additionally.

@TheMMaciek , while `discoverFrom` does filter out known peers, we are joining asynchronously. Additionally, some joinings may fail. So we still have to filter out peers we attempted to join.

> * Another thing if my intuition is right you should use `updatedJoined` to filter out discovered peers and not `joining` if filtering is needed afterall.

@TheMMaciek , please explain.

> * I would also think about maybe moving discovery to the `Joining` class and perform discovery based on the known details in the `joinAll` (the attempted, failed and joined nodes) and only offload the last discovered to the ref that would be available to fetch from.

@TheMMaciek , please elucidate further. I do not understand.

Reply:
> * Another thing is that not all peers in `joining` will always be joined, so the second parameter called `joined` here should rather be called `attempted`.

Changed `joining` to `attempted` and `updatedJoining` to `updatedAttempted`.

> * I would consider if we need another field/parameter for `failed` to join peers - that depends on which road we will choose (see next point)

Joining for each peer will only be attempted once. A `failed` set is not needed.

Reply:
> @TheMMaciek , while discoverFrom does filter out known peers, we are joining asynchronously. Additionally, some joinings may fail. So we still have to filter out peers we attempted to join.

I inspected the discovery functionality again and I think previously discovered peers are filtered out there, so I don't think that the additional filtering is necessary here - correct me if I'm wrong, maybe there is some corner case which I didn't get by briefly checking.

Reply:
Currently the peer discovery is a kind of separate parallel being. I think it was created like this because we wanted to have both tracking of already discovered peers (to not rediscover those again and again on subsequent joins) and to be able to return discovered peers to whomever asks for those. Additionally the joining process before this change didn't keep track of the attempted peers. Right now we have the context of the attempted and joined peers so we could simply offload the discovered not joined peers to the Ref here in the `Joining` class. I think it may get simpler, WDYT @marcinwadon @kpudlik @sdavidp @scasplte2

Reply:
> Joining for each peer will only be attempted once. A failed set is not needed.

Yes they will only be attempted once, but it may be beneficial to keep the information of which joined successfully and which didn't - especially if we would like to run some sanity check at the end of the joining process. Maybe the initial implementation could just log that information, so something like "Joining process finished. Joined X peers and failed to join Y peers ({optionally list of peer ids in a short format})" where X and Y are numbers.

Reply:
All failed joining and peer discovery attempts are now logged. We do need to filter previously discovered peers in addition to the filtering in the peer discovery function because we are joining and discovering from a list of `P2PContext`s in parrallel with `parTraverse`. So, we have to do a second round of filtering just in case.